### ISSUE TITLE
Roughly once per day we get a 500 error because _cache_card does not exist

### ISSUE THREAD

**Issue by bakert** [PennyDreadfulMTG/Penny-Dreadful-Tools#5095]
I assume this is because we delete it before we recreate it and it isn't in a transaction. Hopefully wrapping it in a transaction is easy. Although MySQL doesn't do nested transactions so let's not count our chickens.

```
Failed to execute SELECT * FROM _cache_card AS c WHERE ((1 = 1) AND LOWER(c.name) IN ('corpse cur')) with [] because of (1146, "Table 'cards._cache_card' doesn't exist")

Reported on decksite by logged_out```
--------------------------------------------------------------------------------
Request Method: GET
Path: /seasons/6/cards/Corpse Cur/?
Cookies: {'__cfduid': 'd9b3f9a296fa4a7f0f87c5904ad0423e91532544226', 'session': '.eJwVyrsOwjAMQNF_8YywRKd2AxaGTEit2JCJTZoqLzXuUCH-nbDeez4QsqUgMMBCcIBMm86nZ1XSf6vdXkJHxk_jubcXc727URd53F5sem5eaXWiTc6qpQ6IRVLaeRXi9xYiOW-PNkckjj61mUsQTFmlInx_ZLQrhg.DkBBhg.026ANZyCARN52l8GLibYtijX1d0'}
Endpoint: seasons.card
View Args: {'name': 'Corpse Cur'}
Person: logged_out
Referrer: None
Request Data: {}
Host: pennydreadfulmagic.com
Accept-Encoding: gzip
Cf-Ipcountry: US
X-Forwarded-For: 35.153.255.234, 173.245.52.153
Cf-Ray: 4425c3ee0ca621c8-EWR
X-Forwarded-Proto: https
Cf-Visitor: {"scheme":"https"}
From: crawler.feedback+dc@gmail.com
User-Agent: MauiBot (crawler.feedback+dc@gmail.com)
Cookie: __cfduid=d9b3f9a296fa4a7f0f87c5904ad0423e91532544226; session=.eJwVyrsOwjAMQNF_8YywRKd2AxaGTEit2JCJTZoqLzXuUCH-nbDeez4QsqUgMMBCcIBMm86nZ1XSf6vdXkJHxk_jubcXc727URd53F5sem5eaXWiTc6qpQ6IRVLaeRXi9xYiOW-PNkckjj61mUsQTFmlInx_ZLQrhg.DkBBhg.026ANZyCARN52l8GLibYtijX1d0
Cf-Connecting-Ip: 35.153.255.234
X-Forwarded-Host: pennydreadfulmagic.com
X-Forwarded-Server: pennydreadfulmagic.com
Connection: Keep-Alive

--------------------------------------------------------------------------------
DatabaseException
Stack Trace:
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2309, in call
return self.wsgi_app(environ, start_response)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2295, in wsgi_app
response = self.handle_exception(e)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2292, in wsgi_app
response = self.full_dispatch_request()
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1815, in full_dispatch_request
rv = self.handle_user_exception(e)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1718, in handle_user_exception
reraise(exc_type, exc_value, tb)
File "/home/discord/.local/lib/python3.6/site-packages/flask/_compat.py", line 35, in reraise
raise value
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1813, in full_dispatch_request
rv = self.dispatch_request()
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1799, in dispatch_request
return self.view_functionsrule.endpoint
File "./decksite/cache.py", line 71, in decorated_function
response = make_response(f(*args, **kwargs))
File "./decksite/main.py", line 103, in card
c = cs.load_card(oracle.valid_name(urllib.parse.unquote_plus(name)), season_id=get_season_id())
File "./decksite/data/card.py", line 43, in load_card
c = guarantee.exactly_one(oracle.load_cards([name]))
File "./magic/oracle.py", line 45, in load_cards
rs = db().execute(sql)
File "./shared/database.py", line 50, in execute
raise DatabaseException('Failed to execute {sql} with {args} because of {e}'.format(sql=sql, args=args, e=e))
```

**Comment by bakert** [comment]
```
Failed to execute SELECT * FROM _cache_card AS c WHERE ((1 = 1) AND LOWER(c.name) IN ('herald of dromoka')) with [] because of (1146, "Table 'cards._cache_card' doesn't exist")

Reported on decksite by logged_out```
--------------------------------------------------------------------------------
Request Method: GET
Path: /cards/Herald of Dromoka/?locale=es
Cookies: {'__cfduid': 'd74fc93a61ee7c4fe5d4716c2e791829b1532888784', 'session': 'eyJsb2NhbGUiOiJlbiJ9.Dj-YvQ.e5AGH_aiqDLNkn_D1diSwnyS8R8'}
Endpoint: card
View Args: {'name': 'Herald of Dromoka'}
Person: logged_out
Referrer: None
Request Data: {}
Host: pennydreadfulmagic.com
Accept-Encoding: gzip
Cf-Ipcountry: DE
X-Forwarded-For: 2a01:4f8:150:2314:0:0:0:2, 162.158.92.12
Cf-Ray: 4421a513dc4c96b2-FRA
X-Forwarded-Proto: https
Cf-Visitor: {"scheme":"https"}
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,/;q=0.8
User-Agent: Mozilla/5.0 (compatible; BLEXBot/1.0; +http://webmeup-crawler.com/)
Cookie: __cfduid=d74fc93a61ee7c4fe5d4716c2e791829b1532888784; session=eyJsb2NhbGUiOiJlbiJ9.Dj-YvQ.e5AGH_aiqDLNkn_D1diSwnyS8R8
Cf-Connecting-Ip: 2a01:4f8:150:2314:0:0:0:2
X-Forwarded-Host: pennydreadfulmagic.com
X-Forwarded-Server: pennydreadfulmagic.com
Connection: Keep-Alive

--------------------------------------------------------------------------------
DatabaseException
Stack Trace:
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2309, in call
return self.wsgi_app(environ, start_response)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2295, in wsgi_app
response = self.handle_exception(e)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2292, in wsgi_app
response = self.full_dispatch_request()
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1815, in full_dispatch_request
rv = self.handle_user_exception(e)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1718, in handle_user_exception
reraise(exc_type, exc_value, tb)
File "/home/discord/.local/lib/python3.6/site-packages/flask/_compat.py", line 35, in reraise
raise value
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1813, in full_dispatch_request
rv = self.dispatch_request()
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1799, in dispatch_request
return self.view_functionsrule.endpoint
File "./decksite/cache.py", line 71, in decorated_function
response = make_response(f(*args, **kwargs))
File "./decksite/main.py", line 103, in card
c = cs.load_card(oracle.valid_name(urllib.parse.unquote_plus(name)), season_id=get_season_id())
File "./decksite/data/card.py", line 43, in load_card
c = guarantee.exactly_one(oracle.load_cards([name]))
File "./magic/oracle.py", line 45, in load_cards
rs = db().execute(sql)
File "./shared/database.py", line 50, in execute
raise DatabaseException('Failed to execute {sql} with {args} because of {e}'.format(sql=sql, args=args, e=e))
```

**Comment by bakert** [comment]
```
Failed to execute SELECT * FROM _cache_card AS c WHERE ((1 = 1) AND LOWER(c.name) IN ('celestial kirin')) with [] because of (1146, "Table 'cards._cache_card' doesn't exist")

Reported on decksite by logged_out```
--------------------------------------------------------------------------------
Request Method: GET
Path: /seasons/6/cards/Celestial Kirin/?
Cookies: {}
Endpoint: seasons.card
View Args: {'name': 'Celestial Kirin'}
Person: logged_out
Referrer: None
Request Data: {}
Host: pennydreadfulmagic.com
Accept-Encoding: gzip
Cf-Ipcountry: RU
X-Forwarded-For: 84.201.133.1, 172.69.138.36
Cf-Ray: 440ec3256f0db63b-TLL
X-Forwarded-Proto: https
Cf-Visitor: {"scheme":"https"}
User-Agent: Mozilla/5.0 (compatible; YandexBot/3.0; +http://yandex.com/bots)
From: support@search.yandex.ru
Accept: /
Accept-Language: en, *;q=0.01
Cf-Connecting-Ip: 84.201.133.1
X-Forwarded-Host: pennydreadfulmagic.com
X-Forwarded-Server: pennydreadfulmagic.com
Connection: Keep-Alive

--------------------------------------------------------------------------------
DatabaseException
Stack Trace:
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2309, in call
return self.wsgi_app(environ, start_response)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2295, in wsgi_app
response = self.handle_exception(e)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2292, in wsgi_app
response = self.full_dispatch_request()
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1815, in full_dispatch_request
rv = self.handle_user_exception(e)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1718, in handle_user_exception
reraise(exc_type, exc_value, tb)
File "/home/discord/.local/lib/python3.6/site-packages/flask/_compat.py", line 35, in reraise
raise value
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1813, in full_dispatch_request
rv = self.dispatch_request()
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1799, in dispatch_request
return self.view_functionsrule.endpoint
File "./decksite/cache.py", line 71, in decorated_function
response = make_response(f(*args, **kwargs))
File "./decksite/main.py", line 103, in card
c = cs.load_card(oracle.valid_name(urllib.parse.unquote_plus(name)), season_id=get_season_id())
File "./decksite/data/card.py", line 43, in load_card
c = guarantee.exactly_one(oracle.load_cards([name]))
File "./magic/oracle.py", line 45, in load_cards
rs = db().execute(sql)
File "./shared/database.py", line 50, in execute
raise DatabaseException('Failed to execute {sql} with {args} because of {e}'.format(sql=sql, args=args, e=e))
```

**Comment by bakert** [comment]
```
Failed to execute SELECT * FROM _cache_card AS c WHERE ((1 = 1) AND LOWER(c.name) IN ('avizoa')) with [] because of (1146, "Table 'cards._cache_card' doesn't exist")

Reported on decksite by logged_out```
--------------------------------------------------------------------------------
Request Method: GET
Path: /cards/Avizoa/?locale=es
Cookies: {}
Endpoint: card
View Args: {'name': 'Avizoa'}
Person: logged_out
Referrer: None
Request Data: {}
Host: pennydreadfulmagic.com
Accept-Encoding: gzip
Cf-Ipcountry: US
X-Forwarded-For: 157.55.39.79, 108.162.246.127
Cf-Ray: 44057e0a6c0c7a17-SEA
X-Forwarded-Proto: https
Cf-Visitor: {"scheme":"https"}
Cache-Control: no-cache
Pragma: no-cache
Accept: /
From: bingbot(at)microsoft.com
User-Agent: Mozilla/5.0 (compatible; bingbot/2.0; +http://www.bing.com/bingbot.htm)
Cf-Connecting-Ip: 157.55.39.79
X-Forwarded-Host: pennydreadfulmagic.com
X-Forwarded-Server: pennydreadfulmagic.com
Connection: Keep-Alive

--------------------------------------------------------------------------------
DatabaseException
Stack Trace:
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2309, in call
return self.wsgi_app(environ, start_response)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2295, in wsgi_app
response = self.handle_exception(e)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2292, in wsgi_app
response = self.full_dispatch_request()
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1815, in full_dispatch_request
rv = self.handle_user_exception(e)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1718, in handle_user_exception
reraise(exc_type, exc_value, tb)
File "/home/discord/.local/lib/python3.6/site-packages/flask/_compat.py", line 35, in reraise
raise value
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1813, in full_dispatch_request
rv = self.dispatch_request()
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1799, in dispatch_request
return self.view_functionsrule.endpoint
File "./decksite/cache.py", line 71, in decorated_function
response = make_response(f(*args, **kwargs))
File "./decksite/main.py", line 103, in card
c = cs.load_card(oracle.valid_name(urllib.parse.unquote_plus(name)), season_id=get_season_id())
File "./decksite/data/card.py", line 43, in load_card
c = guarantee.exactly_one(oracle.load_cards([name]))
File "./magic/oracle.py", line 45, in load_cards
rs = db().execute(sql)
File "./shared/database.py", line 50, in execute
raise DatabaseException('Failed to execute {sql} with {args} because of {e}'.format(sql=sql, args=args, e=e))
```

**Comment by bakert** [comment]
Theorectically this is already in a transaction:

```
def update_cache() -> None:
    db().begin()
    db().execute('DROP TABLE IF EXISTS _cache_card')
    db().execute('SET group_concat_max_len=100000')
    db().execute('CREATE TABLE _cache_card AS {base_query}'.format(base_query=base_query()))
    db().execute('CREATE INDEX idx_name_name on _cache_card (name(142))')
    db().commit()
```

**Comment by bakert** [comment]
Our transactions do not work which can be seen executing this code and simultaneously querying the monkey table:

```
@APP.route('/test/')
def test():
    from decksite.database import db
    import time
    db().begin()
    db().execute('DROP TABLE IF EXISTS monkey')
    time.sleep(5)
    db().execute('CREATE TABLE monkey (id INT)')
    db().execute('INSERT INTO monkey (id) VALUES (2)')
    db().commit()
    return 'done'
```

You'll get "table does not exist" during the sleep :/

**Comment by bakert** [comment]
ok looks like mysql doesn't support transactions including a DROP TABLE :/

**Comment by bakert** [comment]
Yeah, our transactions work just fine for two UPDATEs and similar. Grr.

**Comment by bakert** [comment]
we could create the new table and then RENAME it. this messes up foreign keys though, i think.

**Comment by bakert** [comment]
```
Failed to execute SELECT * FROM _cache_card AS c WHERE ((1 = 1) AND LOWER(c.name) IN ('reverent hunter')) with [] because of (1146, "Table 'cards._cache_card' doesn't exist")

Reported on decksite by logged_out```
--------------------------------------------------------------------------------
Request Method: GET
Path: /cards/Reverent Hunter/?locale=en
Cookies: {}
Endpoint: card
View Args: {'name': 'Reverent Hunter'}
Person: logged_out
Referrer: None
Request Data: {}
Host: pennydreadfulmagic.com
Accept-Encoding: gzip
Cf-Ipcountry: US
X-Forwarded-For: 46.229.168.65, 172.68.54.30
Cf-Ray: 4453155fe8125a68-BOS
X-Forwarded-Proto: https
Cf-Visitor: {"scheme":"https"}
Accept: text/html
User-Agent: Mozilla/5.0 (compatible; SemrushBot/2~bl; +http://www.semrush.com/bot.html)
Cf-Connecting-Ip: 46.229.168.65
X-Forwarded-Host: pennydreadfulmagic.com
X-Forwarded-Server: pennydreadfulmagic.com
Connection: Keep-Alive

--------------------------------------------------------------------------------
DatabaseException
Stack Trace:
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2309, in call
return self.wsgi_app(environ, start_response)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2295, in wsgi_app
response = self.handle_exception(e)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2292, in wsgi_app
response = self.full_dispatch_request()
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1815, in full_dispatch_request
rv = self.handle_user_exception(e)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1718, in handle_user_exception
reraise(exc_type, exc_value, tb)
File "/home/discord/.local/lib/python3.6/site-packages/flask/_compat.py", line 35, in reraise
raise value
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1813, in full_dispatch_request
rv = self.dispatch_request()
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1799, in dispatch_request
return self.view_functionsrule.endpoint
File "./decksite/cache.py", line 71, in decorated_function
response = make_response(f(*args, **kwargs))
File "./decksite/main.py", line 103, in card
c = cs.load_card(oracle.valid_name(urllib.parse.unquote_plus(name)), season_id=get_season_id())
File "./decksite/data/card.py", line 43, in load_card
c = guarantee.exactly_one(oracle.load_cards([name]))
File "./magic/oracle.py", line 45, in load_cards
rs = db().execute(sql)
File "./shared/database.py", line 50, in execute
raise DatabaseException('Failed to execute {sql} with {args} because of {e}'.format(sql=sql, args=args, e=e))
```

**Comment by bakert** [comment]
```
Failed to execute SELECT * FROM _cache_card AS c WHERE ((1 = 1) AND LOWER(c.name) IN ('dimir infiltrator')) with [] because of (1146, "Table 'cards._cache_card' doesn't exist")

Reported on decksite by logged_out```
--------------------------------------------------------------------------------
Request Method: GET
Path: /cards/Dimir Infiltrator/?locale=pl
Cookies: {}
Endpoint: card
View Args: {'name': 'Dimir Infiltrator'}
Person: logged_out
Referrer: None
Request Data: {}
Host: pennydreadfulmagic.com
Accept-Encoding: gzip
Cf-Ipcountry: US
X-Forwarded-For: 46.229.168.79, 108.162.219.180
Cf-Ray: 44b90cdf5ffc0ec1-EWR
X-Forwarded-Proto: https
Cf-Visitor: {"scheme":"https"}
Accept: text/html
User-Agent: Mozilla/5.0 (compatible; SemrushBot/2~bl; +http://www.semrush.com/bot.html)
Cf-Connecting-Ip: 46.229.168.79
X-Forwarded-Host: pennydreadfulmagic.com
X-Forwarded-Server: pennydreadfulmagic.com
Connection: Keep-Alive

--------------------------------------------------------------------------------
DatabaseException
Stack Trace:
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2309, in call
return self.wsgi_app(environ, start_response)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2295, in wsgi_app
response = self.handle_exception(e)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2292, in wsgi_app
response = self.full_dispatch_request()
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1815, in full_dispatch_request
rv = self.handle_user_exception(e)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1718, in handle_user_exception
reraise(exc_type, exc_value, tb)
File "/home/discord/.local/lib/python3.6/site-packages/flask/_compat.py", line 35, in reraise
raise value
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1813, in full_dispatch_request
rv = self.dispatch_request()
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1799, in dispatch_request
return self.view_functionsrule.endpoint
File "./decksite/cache.py", line 79, in decorated_function
response = make_response(f(*args, **kwargs))
File "./decksite/main.py", line 103, in card
c = cs.load_card(oracle.valid_name(urllib.parse.unquote_plus(name)), season_id=get_season_id())
File "./decksite/data/card.py", line 43, in load_card
c = guarantee.exactly_one(oracle.load_cards([name]))
File "./magic/oracle.py", line 45, in load_cards
rs = db().execute(sql)
File "./shared/database.py", line 50, in execute
raise DatabaseException('Failed to execute {sql} with {args} because of {e}'.format(sql=sql, args=args, e=e))
```

**Comment by bakert** [comment]
I think what we need to do here is create the table as _cache_card2 or something and then swap it over in a transaction?

**Comment by jgabrielygalan** [comment]
We discussed that solution in the great January crisis. But I can't remember if we decided for it, and simply it hasn't happened yet.

**Comment by bakert** [comment]
```
Failed to execute SELECT * FROM _cache_card AS c WHERE ((1 = 1) AND LOWER(c.name) IN ('cascading cataracts')) with [] because of (1146, "Table 'cards._cache_card' doesn't exist")

Reported on decksite by logged_out```
--------------------------------------------------------------------------------
Request Method: GET
Path: /cards/Cascading Cataracts/?locale=fr
Cookies: {}
Endpoint: card
View Args: {'name': 'Cascading Cataracts'}
Person: logged_out
Referrer: None
Request Data: {}
Host: pennydreadfulmagic.com
Accept-Encoding: gzip
Cf-Ipcountry: DE
X-Forwarded-For: 144.76.137.254, 162.158.111.99
Cf-Ray: 450b72eadc61bf57-AMS
X-Forwarded-Proto: https
Cf-Visitor: {"scheme":"https"}
Accept: text/html,text/plain,text/xml,text/*,application/xml,application/xhtml+xml,application/rss+xml,application/atom+xml,application/rdf+xml,application/php,application/x-php,application/x-httpd-php
User-Agent: Mozilla/5.0 (compatible; MJ12bot/v1.4.8; http://mj12bot.com/)
Accept-Language: en
Cf-Connecting-Ip: 144.76.137.254
X-Forwarded-Host: pennydreadfulmagic.com
X-Forwarded-Server: pennydreadfulmagic.com
Connection: Keep-Alive

--------------------------------------------------------------------------------
DatabaseException
Stack Trace:
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2309, in call
return self.wsgi_app(environ, start_response)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2295, in wsgi_app
response = self.handle_exception(e)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2292, in wsgi_app
response = self.full_dispatch_request()
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1815, in full_dispatch_request
rv = self.handle_user_exception(e)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1718, in handle_user_exception
reraise(exc_type, exc_value, tb)
File "/home/discord/.local/lib/python3.6/site-packages/flask/compat.py", line 35, in reraise
raise value
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1813, in full_dispatch_request
rv = self.dispatch_request()
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1799, in dispatch_request
return self.view_functionsrule.endpoint
File "./decksite/cache.py", line 79, in decorated_function
response = make_response(f(*args, **kwargs))
File "./decksite/main.py", line 106, in card
c = cs.load_card(oracle.valid_name(urllib.parse.unquote_plus(name)), season_id=get_season_id())
File "./decksite/data/card.py", line 43, in load_card
c = guarantee.exactly_one(oracle.load_cards([name]))
File "./magic/oracle.py", line 45, in load_cards
rs = db().select(sql)
File "./shared/database.py", line 39, in select
[, rows] = self.execute_anything(sql, args)
File "./shared/database.py", line 58, in execute_anything
raise DatabaseException('Failed to execute {sql} with {args} because of {e}'.format(sql=sql, args=args, e=e)
```

**Comment by bakert** [comment]
```
Failed to execute SELECT * FROM _cache_card AS c WHERE ((1 = 1) AND LOWER(c.name) IN ('ascendant evincar')) with [] because of (1146, "Table 'cards._cache_card' doesn't exist")

Reported on decksite by logged_out```
--------------------------------------------------------------------------------
Request Method: GET
Path: /cards/Ascendant Evincar/?locale=pt_BR
Cookies: {'__cfduid': 'dc4d35a2b1cfdc8d73b7967b7f18690691535585257', 'session': 'eyJsb2NhbGUiOiJqYSJ9.Dmi9kg.zEdm4c4CPS3uBaBQazGe2jAcQGI'}
Endpoint: card
View Args: {'name': 'Ascendant Evincar'}
Person: logged_out
Referrer: None
Request Data: {}
Host: pennydreadfulmagic.com
Accept-Encoding: gzip
Cf-Ipcountry: DE
X-Forwarded-For: 2a01:4f8:150:2314:0:0:0:2, 162.158.93.161
Cf-Ray: 4522cb6bce0626de-FRA
X-Forwarded-Proto: https
Cf-Visitor: {"scheme":"https"}
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,/;q=0.8
User-Agent: Mozilla/5.0 (compatible; BLEXBot/1.0; +http://webmeup-crawler.com/)
Cookie: __cfduid=dc4d35a2b1cfdc8d73b7967b7f18690691535585257; session=eyJsb2NhbGUiOiJqYSJ9.Dmi9kg.zEdm4c4CPS3uBaBQazGe2jAcQGI
Cf-Connecting-Ip: 2a01:4f8:150:2314:0:0:0:2
X-Forwarded-Host: pennydreadfulmagic.com
X-Forwarded-Server: pennydreadfulmagic.com
Connection: Keep-Alive

--------------------------------------------------------------------------------
DatabaseException
Stack Trace:
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2309, in call
return self.wsgi_app(environ, start_response)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2295, in wsgi_app
response = self.handle_exception(e)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 2292, in wsgi_app
response = self.full_dispatch_request()
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1815, in full_dispatch_request
rv = self.handle_user_exception(e)
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1718, in handle_user_exception
reraise(exc_type, exc_value, tb)
File "/home/discord/.local/lib/python3.6/site-packages/flask/compat.py", line 35, in reraise
raise value
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1813, in full_dispatch_request
rv = self.dispatch_request()
File "/home/discord/.local/lib/python3.6/site-packages/flask/app.py", line 1799, in dispatch_request
return self.view_functionsrule.endpoint
File "./decksite/cache.py", line 79, in decorated_function
response = make_response(f(*args, **kwargs))
File "./decksite/main.py", line 106, in card
c = cs.load_card(oracle.valid_name(urllib.parse.unquote_plus(name)), season_id=get_season_id())
File "./decksite/data/card.py", line 43, in load_card
c = guarantee.exactly_one(oracle.load_cards([name]))
File "./magic/oracle.py", line 45, in load_cards
rs = db().select(sql)
File "./shared/database.py", line 39, in select
[, rows] = self.execute_anything(sql, args)
File "./shared/database.py", line 58, in execute_anything
raise DatabaseException('Failed to execute {sql} with {args} because of {e}'.format(sql=sql, args=args, e=e))
```

**Comment by bakert** [comment]
Whatever we do here we should probably do for #5206 and *vice versa*.

**Comment by bakert** [comment]
This isn't quite the same thing as #5206. One is about blowing away the db completely and the other is about dropping and recreating _cache_card only.

**Comment by bakert** [comment]
Reduced the amount of time that _cache_card is not available considerably in a0b2637c.

Doesn't actually make this impossible but the switchover should be an eyeblink instead of several seconds. Good enough!
